apps/emqx_authz/test/emqx_authz_http_SUITE.erl

```diff
@@ -22,75 +23,350 @@
+            }),
+
+    ClientInfo = #{clientid => <<"client id">>,
+                   username => <<"user name">>,
+                   peerhost => {127,0,0,1},
+                   protocol => <<"MQTT">>,
+                   mountpoint => <<"MOUNTPOINT">>,
+                   zone => default,
+                   listener => {tcp, default}
+                  },
+
+    allow = emqx_access_control:authorize(ClientInfo, publish, <<"t">>).
+
+t_path_params(_Config) ->
+    ok = setup_handler_and_config(
+           fun(Req0, State) ->
+                   <<"/authz/"
+                     "username/user%20name/"
+                     "clientid/client%20id/"
+                     "peerhost/127.0.0.1/"
+                     "proto_name/MQTT/"
+                     "mountpoint/MOUNTPOINT/"
+                     "topic/t/"
+                     "action/publish">> = cowboy_req:path(Req0),
+                   Req = cowboy_req:reply(200, Req0),
+                   {ok, Req, State}
+           end,
+           #{<<"path">> => <<"username/${username}/"
+                             "clientid/${clientid}/"
+                             "peerhost/${peerhost}/"
+                             "proto_name/${proto_name}/"
+                             "mountpoint/${mountpoint}/"
+                             "topic/${topic}/"
+                             "action/${action}">>
+            }),
+
+    ClientInfo = #{clientid => <<"client id">>,
+                   username => <<"user name">>,
+                   peerhost => {127,0,0,1},
+                   protocol => <<"MQTT">>,
+                   mountpoint => <<"MOUNTPOINT">>,
+                   zone => default,
+                   listener => {tcp, default}
+                  },
+
+    allow = emqx_access_control:authorize(ClientInfo, publish, <<"t">>).
+
+t_json_body(_Config) ->
+    ok = setup_handler_and_config(
+           fun(Req0, State) ->
+                   <<"/authz/"
+                     "username/user%20name/"
+                     "clientid/client%20id/"
+                     "peerhost/127.0.0.1/"
+                     "proto_name/MQTT/"
+                     "mountpoint/MOUNTPOINT/"
+                     "topic/t/"
+                     "action/publish">> = cowboy_req:path(Req0),
+                   {ok, RawBody, Req1} = cowboy_req:read_body(Req0),
+
+                   #{<<"username">> := <<"user name">>,
+                     <<"clientid">> := <<"client id">>,
+                     <<"peerhost">> := <<"127.0.0.1">>,
+                     <<"proto_name">> := <<"MQTT">>,
+                     <<"mountpoint">> := <<"MOUNTPOINT">>,
+                     <<"topic">> := <<"t">>,
+                     <<"action">> := <<"publish">>} = jiffy:decode(RawBody, [return_maps]),
+                   Req = cowboy_req:reply(200, Req1),
+                   {ok, Req, State}
+           end,
+           #{<<"method">> => <<"post">>,
+             <<"path">> => <<"username/${username}/"
```

Comment:
Curiosity: is it needed to pass the same parameters both as a path and in the JSON body?  Shouldn't the body be enough?

Reply:
>Curiosity: is it needed to pass the same parameters both as a path and in the JSON body? Shouldn't the body be enough?

It's a matter of design, but we may imagine a situation when authz server implements some routing, like
```
POST /authz/users/someuser/authorize

{"topic": "foobar"}
```

Reply:
Ok.  Since it had the same parameters on both path and body, it appeared as if the path was always needed.  :slightly_smiling_face: